### PR TITLE
LTP install both 32bit and 64bit package for QE devel + cleanup

### PR DESCRIPTION
**Pull request by pevik**
* Install also ltp-32bit for QA of the development
* Allow install more than one package in LTP_PKG
* Cleanup: drop `LTP_STABLE`
* Log LTP version with record_info()

Motivation for installing also ltp-32bit for QA of the development
Since b19aa7a99 LTP packages can be installed both 32bit and 64bit on
the same system. Installing both of them helps to avoid having special
job for QEMU: install_ltp+sle+Online-m32, install_ltp_baremetal_m32
(osd) and install_ltp+opensuse+DVD-m32 (o3) when installing from
packages. This speedup mainly baremetal testing, but also saves space
(drop extra QCOW2 image used just for 32bit LTP package).

These jobs for 32bit packages still needs to be used for LTP compilation
from git testing (which is not used for production, but just for
development or LTP prerelease testing).

Installation from git stays the same, installing only single chosen
version (due significant slow down).

Verification run:
* QE devel installing both packages `ltp ltp-32bit`: http://quasar.suse.cz/tests/6480 http://quasar.suse.cz/tests/6481
* QE released product installs `qa_test_ltp` by default: http://quasar.suse.cz/tests/6482#step/install_ltp/49
* `LTP_PKG` defined still works: `ltp ltp-32bit`: http://quasar.suse.cz/tests/6483, `ltp-32bit`: http://quasar.suse.cz/tests/6488
* LTP version with record_info(): install_ltp package: http://quasar.suse.cz/tests/6483#step/install_ltp/49 http://quasar.suse.cz/tests/6483#step/install_ltp/61 (both versions),  http://quasar.suse.cz/tests/6490#step/install_ltp/49 http://quasar.suse.cz/tests/6485#step/install_ltp/51 (single version jeOS), install_ltp git: http://quasar.suse.cz/tests/6489#step/boot_ltp/69 http://quasar.suse.cz/tests/6486#step/install_ltp/182 http://quasar.suse.cz/tests/6487#step/install_ltp/196

Done: https://build.suse.de/request/show/239213 https://build.opensuse.org/request/show/884059 https://github.com/SUSE/qa-testsuites/pull/950

### CODE REVIEW

File: tests/kernel/install_ltp.pm [hunk 1]
```diff
@@ -224,12 +224,24 @@ sub add_ltp_repo {
...
+sub get_default_pkg {
+    return (want_stable && is_sle) ? 'qa_test_ltp' : 'ltp';
+}
 
-    zypper_call("in --recommends $pkg");
-    script_run "rpm -qi $pkg | tee /opt/ltp_version";
-    assert_script_run "find " . get_ltproot() . q(/testcases/bin/openposix/conformance/interfaces/ -name '*.run-test' > ~/openposix-test-list);
+sub install_from_repo {
+    my $default_pkg      = get_default_pkg;
+    my @pkgs             = (get_var('LTP_PKG', $default_pkg));
```
Comment (cfconrad) [inline tests/kernel/install_ltp.pm:233]: what about, so we can specify both packages in that variable
```suggestion
    my @pkgs             = split(/\s*,\s*/, get_var('LTP_PKG', $default_pkg));
```

I didn't checked other usage of `LTP_PKG`

File: tests/kernel/install_ltp.pm [hunk 2]
```diff
@@ -224,12 +224,24 @@ sub add_ltp_repo {
...
+    my $install_both_pkg = $pkgs[0] eq 'both';
+    my $want_32bit       = undef;
+
+    @pkgs = ($default_pkg, $default_pkg . '-32bit') if ($install_both_pkg);
+
+    for my $pkg (@pkgs) {
+        $want_32bit = $pkg =~ m/32bit/ if ($install_both_pkg);
```
Comment (cfconrad) [inline tests/kernel/install_ltp.pm:240]: why do you need this extra check, could there be a case where the package contain `32bit` but you wanna run it in 64bit ?
```suggestion
        $want_32bit = $pkg =~ m/32bit/;
```

File: tests/kernel/install_ltp.pm [hunk 3]
```diff
@@ -224,12 +224,24 @@ sub add_ltp_repo {
...
+        zypper_call("in --recommends $pkg");
+        script_run "rpm -qi $pkg | tee /opt/ltp_version";
```
Comment (cfconrad) [inline tests/kernel/install_ltp.pm:242]: I like to see this in the tests... 
```suggestion
        record_info('LTP_VERSION', script_output("rpm -qi $pkg | tee /opt/ltp_version"));
```

**Review by cfconrad — APPROVED** [review]
Looks good to me, just some comments which might be useful.

File: tests/kernel/install_ltp.pm [hunk 4]
```diff
@@ -224,12 +224,24 @@ sub add_ltp_repo {
     zypper_ar($repo, name => 'ltp_repo');
 }
 
-sub install_from_repo {
-    my $pkg = get_var('LTP_PKG', (want_stable && is_sle) ? 'qa_test_ltp' : 'ltp');
+sub get_default_pkg {
+    return (want_stable && is_sle) ? 'qa_test_ltp' : 'ltp';
```
Comment (b10n1k) [inline tests/kernel/install_ltp.pm:228]: it doesnt seem you need `is_sle`. `want_stable` default contains `is_sle`. i havent found an example with LTP_STABLE to get the idea when the variable is set but it is not sle. Can you provide to me such example?

File: tests/kernel/install_ltp.pm (hunk 2 above)
Comment (pevik) [reply tests/kernel/install_ltp.pm:240]: Well, I could create `pkgs` as hash with package name and prefix keys. And when I'm adding 32bit package I could define also path. Then the extra check would not be needed.
But I should probably define loop outside of the function to simplify things.

File: tests/kernel/install_ltp.pm (hunk 4 above)
Comment (pevik) [reply tests/kernel/install_ltp.pm:228]: `is_sle` and `want_stable` is not relevant to this PR. Believe me it's needed (setup for QAM tests; not everything is documented). It'd help me more if you could test it for cloud (or tell me how I can do it), but that can wait till final code.

File: tests/kernel/install_ltp.pm (hunk 3 above)
Comment (pevik) [reply tests/kernel/install_ltp.pm:242]: That'd be certainly useful, feel free to implement it.

File: tests/kernel/install_ltp.pm (hunk 1 above)
Comment (pevik) [reply tests/kernel/install_ltp.pm:233]: Well, this sounds better (I was thinking about it myself): instead of `LTP_PKG=both` which sounds confusing, just allow to define more packages: `LTP_PKG=ltp ltp-32bit`. @richiejp any opinion about this?

File: tests/kernel/install_ltp.pm (hunk 4 above)
Comment (pevik) [reply tests/kernel/install_ltp.pm:228]: @b10n1k `want_stable` can be used on o3. But stable LTP on o3 is also called `ltp`. 
`qa_test_ltp` is 1) only on osd 2) is stable. That's why double check (for me it was kind of documentation that's why I had no problem to use it twice).

Comment (pevik) [reply tests/kernel/install_ltp.pm:228]: @mloviska FYI

Comment (pevik) [reply tests/kernel/install_ltp.pm:228]: @mdoucha FYI

Comment (b10n1k) [reply tests/kernel/install_ltp.pm:228]: ok. just to make clear. can LTP_STABLE be ever true on o3? 

i agree that it is not relevant to this PR and thats why we can close this conversation. But if it a small glitch it would be also an easy fix as you touch that particular code. or i would like to know for the future. no blocking issue

Comment (pevik) [reply tests/kernel/install_ltp.pm:228]: It should not do anything since https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12203. Regardless setting it you always get `ltp` package with master branch. I'm open to suggestions what to do with the variable, or just send PR. 
It's just instead of tuning LTP openQA runner we prefer to work on LTP itself (openQA is just a runner, tests are what brings a benefit; unless some significant improvement like speedup). But I'll try to have look and improve openQA LTP docs.

File: tests/kernel/install_ltp.pm (hunk 2 above)
Comment (pevik) [reply tests/kernel/install_ltp.pm:240]: @cfconrad Whole thing is working, package fixed. I wonder if I should rewrite this bit, or just ignore it and merge.

**Comment by mdoucha** [comment]
Hi, I don't see any reason why complicate things with `LTP_PKG`. Just install both packages if the arch is `x86_64`.

**Comment by pevik** [comment]
> Hi, I don't see any reason why complicate things with `LTP_PKG`. Just install both packages if the arch is `x86_64`.

Well, it could be on arch is `x86_64` && ! is_released. Because released products does not install it. If you don't miss the flexibility, why not. And I guess nobody tested `qa_test_ltp-32bit` for ages, thus I would not install it when not required. Also, currently `qa_test_ltp` is not synced with git repository, thus it'd fail trying to install it.

File: tests/kernel/install_ltp.pm (hunk 3 above)
Comment (pevik) [reply tests/kernel/install_ltp.pm:242]: BTW this should be not just in `install_ltp.pm`, but also in `schedule_tests()`, which prints it during test run.

**Comment by pevik** [comment]
OK, I'll simplify things: 1) drop `LTP_STABLE` as nobody is using it 2) create default according to current state (including installing both 32bit and 64bit if arch is x86_64 && ! is_released), 3) allow to overwrite the default with `LTP_PKG` (also allow to define more packages).

File: tests/kernel/install_ltp.pm (hunk 4 above)
Comment (b10n1k) [reply tests/kernel/install_ltp.pm]: thanks @pevik

**Review by b10n1k — APPROVED** [review]
+1

**Comment by richiejp** [comment]
LGTM

**Comment by pevik** [comment]
@mdoucha @richiejp @mloviska Updated also examples of our config.

**Comment by pevik** [comment]
Hopefully final version pushed. Changed second commit (split) in `install_from_repo()`. Also install all possible packages at once, just `record_info` and `find` are being done separately (there could also be just single `record_info` and separate packages with blank line).

**Review by czerw — APPROVED** [review]

File: lib/LTP/utils.pm [hunk 5]
```diff
@@ -48,7 +49,18 @@ sub shutdown_ltp {
 }
 
 sub get_ltproot {
-    return get_required_var('TEST_SUITE_NAME') =~ m/[-_]m32$/ ? '/opt/ltp-32' : '/opt/ltp';
+    # TEST_SUITE_NAME is for running 32bit tests (e.g. ltp_syscalls_m32),
+    # checking LTP_PKG is for install_ltp.pm which also uses prepare_ltp_env()
+    my $want_32bit = shift // (get_required_var('TEST_SUITE_NAME') =~ m/[-_]m32$/
+          || get_var('LTP_PKG', '') =~ m/^(ltp|qa_test_ltp)-32bit$/);
+
```
Comment (pevik) [inline lib/LTP/utils.pm:56]: Slightly modified `get_ltproot()` here in the second commit to make sure install_ltp does not fail trying to `cd /opt/ltp/...` when installing only 32bit package: http://quasar.suse.cz/tests/6469#step/install_ltp/230 (corner case, normally it shold not happen.

**Comment by pevik** [comment]
@mdoucha I don't run twice `assert_script_run('generate_lvm_runfile.sh')` in `install_ltp.pm`. Do we want to  be precise and run it also for 32bit built or we don't bother?

File: tests/kernel/install_ltp.pm (hunk 3 above)
Comment (cfconrad) [reply tests/kernel/install_ltp.pm]: I think you already did, nice!

**Comment by mdoucha** [comment]
> @mdoucha I don't run twice `assert_script_run('generate_lvm_runfile.sh')` in `install_ltp.pm`. Do we want to be precise and run it also for 32bit built or we don't bother? (it's not needed for 32bit build, but it's _wrong_ not to run it).
> Or wouldn't be easier just to move calling `generate_lvm_runfile.sh` into `run_ltp.pm`?

Moving it to `run_ltp.pm` won't work because the runfile needs to exist when `boot_ltp` (or `install_ltp` in install+run job) schedules all the `run_ltp` modules. But there's no point running LVM tests in compat mode. Maybe add a comment that we don't care about 32bit LVM tests enough to generate the runfile but otherwise you can ignore it.

File: tests/kernel/install_ltp.pm [hunk 6]
```diff
@@ -224,12 +221,27 @@ sub add_ltp_repo {
     zypper_ar($repo, name => 'ltp_repo');
 }
 
+sub get_default_pkg {
+    if (is_sle) {
+        return 'qa_test_ltp' if (is_released);
```
Comment (mdoucha) [inline tests/kernel/install_ltp.pm:226]: The default for released SLES should be `qa_test_ltp qa_test_ltp-32bit`. I'll probably add 32bit syscalls and cves to the schedule soon.

Comment (pevik) [reply tests/kernel/install_ltp.pm:226]: Sure, I'll fix it.

**Comment by pevik** [comment]
> Moving it to `run_ltp.pm` won't work because the runfile needs to exist when `boot_ltp` (or `install_ltp` in install+run job) schedules all the `run_ltp` modules. But there's no point running LVM tests in compat mode. Maybe add a comment that we don't care about 32bit LVM tests enough to generate the runfile but otherwise you can ignore it.

It wouldn't be too hard to move handling `LTP_PKG` (`@pkgs`) out of `install_from_repo()` to `run()` to make sure `generate_lvm_runfile.sh` is installed on any package. But as `generate_lvm_runfile.sh` currently runs in 64bit, I'll just add explanation that we don't care about 32bit LVM.

**Comment by czerw** [comment]
ltp_openposix https://openqa.suse.de/tests/5821160 has problems, as 64 bit version runs 32bit binaries due to writing runfile into /root.

**Comment by pevik** [comment]
`ltp_openposix` should be fixed now with installing correct version of `libgcc_s1`, but I'll retest it to be sure.

**Comment by pevik** [comment]
I see you already retest it.
